src/collectors/statsd/statsdCollector.py

```diff
@@ -0,0 +1,144 @@
+"""
+The StatsdCollector is capable of receiving data as sent by any statsd client.
+
+This collector was created in order to be able to collect statsd metrics without having to use statsd. Currently, this
+collector supports a subset of statsd metrics as defined in
+https://github.com/etsy/statsd/blob/master/docs/metric_types.md.
+
+In particular, this collector does not support:
+ - sets
+ - gauges with -/+ modifiers
+
+This collector starts a UDP server in a separate thread to receive and parse data. The server puts the data on a
+queue and waits for the collect() method to pull.
+
+"""
+
+import diamond.metric
+import threading
+import socket
+import Queue
+
+ALIVE = True
+valid_types = ["g", "ms"]
+
+
+def parse_metric(raw_metrics):
+    for raw_metric in raw_metrics.split("\n"):
+        data, metric_type = raw_metric.split("|")[:2]
+        metric_name, value = data.split(":")
+        metric = NewMetric(metric_name, value)
+        if metric_type == "c":
+            metric.isCounter = True
+        elif metric_type not in valid_types:
+            raise ValueError("Metric type %i not recognized/supported" % metric_type)
+        yield metric
+
+
+class StatsdCollector(diamond.collector.Collector):
+
+    def __init__(self, *args, **kwargs):
+        super(StatsdCollector, self).__init__(*args, **kwargs)
+        self.listener_thread = None
+
+    def get_default_config(self):
+        """
+        Returns default collector settings
+        """
+        config = super(StatsdCollector, self).get_default_config()
+        config.update({
+            'listener_host': '127.0.0.1',
+            'listener_port': 8787,
+            'path': 'statsd',
+        })
+        return config
+
+    def collect(self):
+        if not self.listener_thread:
+            self.start_listener()
+
+        while True:
+            try:
+                data = self.listener_thread.queue.get(False)
+                if data.isCounter:
+                    self.publish(data.path, data.value, metric_type='COUNTER')
+                else:
+                    self.publish(data.path, data.value)
+            except Queue.Empty:
+                self.log.info('Queue is empty')
+                break
+            except Exception as e:
+                self.log.error('type={}, exception={}'.format(type(e), e))
+            self.log.info('Path: {}, Value: {}'.format(data.path, data.value))
+
+    def start_listener(self):
+        self.listener_thread = ListenerThread(self.config['listener_host'],
+                                              self.config['listener_port'],
+                                              self.log)
+        self.listener_thread.start()
+
+    def stop_listener(self):
+        global ALIVE
+        ALIVE = False
+        self.listener_thread.join()
```

Comment:
please ensure that we drain the queue and publish all the events before a shutdown.

Reply:
You will need to merge the thread back in before you empty the queue.